src/persistQueryClient/PersistQueryClientProvider.tsx

```diff
@@ -0,0 +1,46 @@
+import React from 'react'
+
+import { persistQueryClient, PersistQueryClientOptions } from './persist'
+import { QueryClientProvider, QueryClientProviderProps } from '../reactjs'
+import { IsHydratingProvider } from '../reactjs/Hydrate'
+
+export interface PersistQueryClientProviderProps
+  extends QueryClientProviderProps {
+  persistOptions: Omit<PersistQueryClientOptions, 'queryClient'>
+  onSuccess?: () => void
+}
+
+export const PersistQueryClientProvider = ({
+  client,
+  children,
+  persistOptions,
+  onSuccess,
+  ...props
+}: PersistQueryClientProviderProps): JSX.Element => {
+  const [isHydrating, setIsHydrating] = React.useState(true)
+  const refs = React.useRef({ persistOptions, onSuccess })
+
+  React.useEffect(() => {
+    refs.current = { persistOptions, onSuccess }
+  })
+
+  React.useEffect(() => {
+    const [unsubscribe, promise] = persistQueryClient({
+      ...refs.current.persistOptions,
+      queryClient: client,
+    })
+
+    promise.then(() => {
+      refs.current.onSuccess?.()
+      setIsHydrating(false)
+    })
+
+    return unsubscribe
```

Comment:
Bit of an edge case and therefor a nit, but what would currently happen if `client` changes before the restore has happened? (I _think_ it would currently still restore the data from the old one, even if we have unsubscribed?) What should happen? Might make a good test.

I'm also wondering about the `promise.then` above, will that (and therefor `onSuccess`) get called even if we have done the effect cleanup? Should it?

Reply:
yes, it would still restore from the old one and also try to restore from the new one. So there might be race conditions. I don't think the client is meant to change at this level, but it can. Some possible solutions:

- move `client` to the `refs` and change the dependency array to an empty array. This communicates that we only want to restore once, but it would not work if the client really changes.
- keep as-is, because race conditions wouldn't really matter. `hydration` can be done multiple times, even from different sources. The query with the highest `dataUpdatedAt` time "wins". So it doesn't matter which one gets hydrated first.
- somehow defer the second call until the first promise has finished - maybe by storing the promise on a ref?

I would think that keep as-is should be good enough. One related problem is that `promise.then` would set `isHydrating` to `false`. I think we need to set it to `true` when the effect starts, or maybe even keep a count of currently running hydrations?
I'm also missing error handling, so if hydration promise rejects, we'll stay in `isHydrating` forever. That I definitely need to fix!

Reply:
just realised: the client is the target, not the source of hydration. For the source to change, it would be a different persister. I think that's very unlikely to be the case. For a different client, I can see the case:

- you create one client, start restoring
- then you change the client e.g. because you change something in the `defaultOptions`, maybe in a `useEffect` - you still want to restore and subscribe I guess?

In that case, order really doesn't matter. However, we'd to somehow "wait" until both hydrations are done before updating the Provider 🤔

Reply:
okay so the count is tricky because if we start at `0` and increment in the effect, we get one render cycle with `isHydrating: false`, which is not good.

I tried to do nr. 3 from above with chained promises, storing the previous one on a ref and added a test. Honestly not sure if that code does what I think it does :/

507edca1

Reply:
Hmm, I see what you mean, it's a bit hard to reason about.

I guess if the client changes, we don't really care if the old client was hydrated or not (it's fine either way), we only care that _when_ the new client has been hydrated, then and only then do we call onSuccess and set isHydrating to false?

Just thinking out loud, but maybe you could leverage a `isStale` variable? Something like:

```jsx
  React.useEffect(() => {
    let isStale = false
    setIsHydrating(true)
    const [unsubscribe, promise] = persistQueryClient({
      ...refs.current.persistOptions,
      queryClient: client,
    })

    promise.then(() => {
      if (!isStale) {
        refs.current.onSuccess?.()
        setIsHydrating(false)
      }
    })

    return () => {
      isStale = true
      unsubscribe()
    }
}, [client])
```

Reply:
i committed the change, but now the test I made fails that should check that onSuccess is only called once. I somehow can't run tests locally because of some weird errors. Can we keep the previous implementation, or do you know why onSuccess would be called twice with this implementation if the client changes?

Reply:
I fixed my tests issue. I misread the test results. With the new implementation, `onSuccess` will only be called once, and it was called (and asserted in the test) twice before.

I like your proposal, so I updated the test and it should be green now: 433fd8a2

Reply:
Nice! Looks good to me. :)